src/checkpoint/dispatch/vrt/object_registry.h

```diff
@@ -0,0 +1,150 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#if !defined INCLUDED_CHECKPOINT_DISPATCH_VRT_OBJECT_REGISTRY_H
+#define INCLUDED_CHECKPOINT_DISPATCH_VRT_OBJECT_REGISTRY_H
+
+#include "checkpoint/common.h"
+#include "checkpoint/dispatch/vrt/registry_common.h"
+#include "checkpoint/dispatch/reconstructor.h"
+
+#include <functional>
+#include <vector>
+#include <tuple>
+
+namespace checkpoint { namespace dispatch { namespace vrt {
+namespace objregistry {
+
+template <typename T>
+struct ObjectEntry {
+
+  template <typename Allocator, typename Constructor>
+  ObjectEntry(
+    TypeIdx in_idx,
+    std::size_t in_size,
+    Allocator&& in_allocator,
+    Constructor&& in_constructor
+  ) : idx_(in_idx),
+      size_(in_size),
+      allocator_(in_allocator),
+      constructor_(in_constructor)
+  { }
+
+  TypeIdx idx_ = no_type_idx;              /**< The type index for this ObjT */
+  std::size_t size_ = 0;                   /**< The registered object size */
+  std::function<void*(void)> allocator_;   /**< Do standard allocation for object */
+  std::function<T*(void*)> constructor_;   /**< Construct object on memory */
+};
+
+template <typename T>
+using RegistryType = std::vector<ObjectEntry<T>>;
+
+template <typename ObjT>
+struct Registrar {
+  Registrar();
+  TypeIdx index;
+};
+
+template <typename ObjT>
+struct Type {
+  static TypeIdx const idx;
+};
+
+template <typename T>
+inline RegistryType<T>& getRegistry() {
+  static RegistryType<T> reg;
+  return reg;
+}
+
+template <typename ObjT>
+Registrar<ObjT>::Registrar() {
+  using BaseType = typename ObjT::_CheckpointVirtualSerializerBaseType;
+
+  auto& reg = getRegistry<BaseType>();
+  index = static_cast<TypeIdx>(reg.size());
+
+  debug_checkpoint("object registrar: %zu, %s\n", reg.size(), typeid(ObjT).name());
+
+  reg.emplace_back(
+    ObjectEntry<BaseType>{
+      index,
+      sizeof(ObjT),
+      []()          -> void*       { return std::allocator<ObjT>{}.allocate(1); },
+      [](void* buf) -> BaseType*   { return dispatch::Reconstructor<ObjT>::constructAllowFail(buf); }
+    }
+  );
+}
+
+template <typename ObjT>
+TypeIdx const Type<ObjT>::idx = Registrar<ObjT>().index;
+
+template <typename T>
+inline auto getObjIdx(TypeIdx han) {
+  debug_checkpoint("getObjIdx: han=%d, size=%ld\n", han, getRegistry<T>().size());
+  return getRegistry<T>().at(han).idx_;
+}
+
+template <typename T>
+inline auto getObjSize(TypeIdx han) {
```

Comment:
I feel silly asking this now, but is `getObjSize` actually used anywhere? Can the size field go away entirely? Is there something we anticipate needing it for if not presently?

Reply:
Yes, maybe you missed my earlier comment, but I anticipate needing it for the future when we want to give the user more control over allocation. Currently, I don't have an overload for that in `allocateConstructForPointer`, but it could be easily adding with a callback.